# Introduction to single cell RNA-seq (scRNA-seq) data analysis
Authors: Krithika Bhuvaneshwar, Yuriy Gusev
Affiliation: Innovation Center For Biomedical Informatics (ICBI), and Biomedical Informatics Shared Resource (BISR) at Georgetown University Medical Center (GUMC)

***More about our research work:***
* *ICBI: https://icbi.georgetown.edu*
* *BISR: https://icbi.georgetown.edu/bisr/ and https://lombardi.georgetown.edu/research/sharedresources/bbsr/*

## What is single cell RNA-seq data ?
* Traditional next generation sequencing (NGS) is done on a bulk scale. It is called bulk NGS, and the expression data is averaged
* In contrast scRNA-seq allows users to get thousands of measurements from individual cells across thousands of dimensions (genes)

<img src="https://github.com/ICBI/BISR_Tutorials/blob/main/scRNA-seq/Picture1.png" width="600" height="400">

## Why do we need single cell sequencing ? You can get the same clusters as in flow cytometry
In single cell sequencing, users can 
* sequence every cell 
* look at the transcriptional profiles of each of these clusters to identify different cell types - can get information about each clone 
* Clusters can be further broken down and analyzed by the genes
* Take measurements on 20,000 genes that might be present in samples
* Explore individual clusters - it is possible to re-cluster and get more separation within one cluster 

In contrast flow cytometry and mass cytometry can 
* only sequence a few limited markers 

## Advantages 
* Very high dimensional read out compared to other methods 

## Disadvantages 
* Cannot examine mutations
 
## Applications
* Cancer
* Immunology
* Allowed development of new analysis methods and computational approaches - pseudotime cell ordering i.e. ordering single cells in pseudotime and placing them along a trajectory corresponding to a biological process

## Different technologies
* Earliest - plate based SMART-seq
* Drop seq - more parallel , droplet based, less expensive
* 10 X genomics - latest (similar to Dropset)
* Microwells - parallel single seq without droplets 

## How get the gene expression count matrix file from raw FASTQC files
* Nextflow: https://nf-co.re/scrnaseq which includes the tools in the Figure
* Galaxy: https://training.galaxyproject.org/training-material/topics/single-cell/tutorials/scrna-case_alevin/tutorial.html

<img src="https://github.com/ICBI/BISR_Tutorials/blob/main/scRNA-seq/scrnaseq_pipeline_v1.0_metro_clean.png" width="500" height="200">


**Tools for this processing**
* STARsolo - a droplet-based scRNA-seq-specific variant of the popular genome alignment method STAR. Produces results very close to those of Cellranger 
* Kallisto/ bustools - developed by the originators of the transcriptome quantification method, Kallisto.
* Alevin - another transcriptome analysis method developed by the authors of the Salmon tool.
* Cellranger - Tool from 10X Genomics that uses STAR under the hood

## Popular software tools for analysis of scRNA-seq data matrix files
(a) Monocle 
* Comprehensive package 
* Clustering
* Differential gene expression testing
* Main aseet - Introduced concept of Pseudotime cell ordering

(b) Seurat
* Performs quality control (QC), analysis, and exploration 
* Unsupervised clustering and discovery of cell types and states, 
* Spatial reconstruction of single cell data
* Integrated analysis of single cell RNA-seq across conditions, technologies, and species
* Main asset - ability to use data from different sequencing technology, species or condition, and integrate them to identify shared population across data sets
* Cannot do Pseudotime cell ordering

## Tools for Downstream analysis (regulatory networks)
* SCENIC: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5937676/
* SCA - IRCM: https://github.com/SCA-IRCM (Allows single cell RNA seq data analysis, clustering, creates internal network and infers cell-cell interactions)

## Recommended reading / workshops
* UCLa workshops https://qcb.ucla.edu/collaboratory/workshops/w20-single-cell-rna-seq/
* Single cell RNA sequencing, an e-book generated by the Bioinformatics team at NYU for mastering NGS analysis, including scRNA-seq analysis
https://learn.gencore.bio.nyu.edu/single-cell-rnaseq/
* UC Davis Bioinformatics Core: https://ucdavis-bioinformatics-training.github.io/2021-August-Single-Cell-RNA-Seq-Analysis/
* Broad Institute single cell workshop https://broadinstitute.github.io/2020_scWorkshop/
* Orchestrating Single-Cell Analysis with Bioconductor, an e-book that teaches users some common workflows for the analysis of scRNA-seq http://bioconductor.org/books/release/OSCA/

## Platforms that offer pipelines for single cell RNA-seq analysis  
* Single cell RNA-seq data analyis on the Galaxy platform: https://training.galaxyproject.org/training-material/topics/single-cell/
* Single-cell RNA-ses analysis pipelines on the Seven Bridges Cancer Genomics Cloud platform: https://cgc.sbgenomics.com
* Single cell RNA-seq pipelines using Nextflow: https://nf-co.re/scrnaseq

## References
* https://www.youtube.com/watch?v=k9VFNLLQP8c
* https://medium.com/@HeleneOMICtools/your-top-3-single-cell-rna-sequencing-analysis-tools-221b65fbc57e
* https://training.galaxyproject.org/training-material/topics/single-cell/tutorials/scrna-case_alevin/tutorial.html
* https://training.galaxyproject.org/training-material/topics/single-cell/tutorials/scrna-case_monocle3-rstudio/tutorial.html#monocle-workflow




