{"cells":[{"cell_type":"markdown","metadata":{"id":"9ltdwXvN0dio"},"source":["# scRNA-seq Monocle Tutorial 2 - Trajectories\n","\n","Authors: Krithika Bhuvaneshwar, Yuriy Gusev\n","\n","Affiliation: Innovation Center For Biomedical Informatics (ICBI), and Biomedical Informatics Shared Resource (BISR) at Georgetown University Medical Center (GUMC)\n","\n","***More about our research work:***\n","* *ICBI: https://icbi.georgetown.edu*\n","* *BISR: https://icbi.georgetown.edu/bisr/ and https://lombardi.georgetown.edu/research/sharedresources/bbsr/*\n"]},{"cell_type":"markdown","metadata":{"id":"9KMphPwYFegF"},"source":["##  Before we start\n","\n","If using google colab, change the colab runtype environment to R (default is python). Go to  Runtime -> Change runtime type -> In the \"Notebook settings\", change environment to \"R\""]},{"cell_type":"markdown","metadata":{"id":"JxM7MTI0GD9U"},"source":["## Monocle software\n","Monocle software: https://cole-trapnell-lab.github.io/monocle3/docs/installation/\n","\n","* Monocle 3 is designed for use with absolute transcript counts (e.g. from UMI experiments).\n","* Monocle 3 works \"out-of-the-box\" with the transcript count matrices produced by Cell Ranger, the software pipeline for analyzing experiments from the 10X Genomics Chromium instrument.\n","* Monocle 3 also works well with data from other RNA-Seq workflows such as sci-RNA-Seq and instruments like the Biorad ddSEQ.\n","\n","Reference:\n","Reference: https://cole-trapnell-lab.github.io/monocle3/docs/trajectories/\n"]},{"cell_type":"markdown","source":[],"metadata":{"id":"OBjgg636V4fP"}},{"cell_type":"markdown","metadata":{"id":"_Bur2rTaE5gb"},"source":["### Install monocle package - this can take a SEVERAL minutes\n","Note - if you encounter errors during installaton, please refer this page: https://cole-trapnell-lab.github.io/monocle3/docs/installation/"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"rpgabL0TEXdT"},"outputs":[],"source":["#install monocle3 through the cole-trapnell-lab GitHub\n","install.packages(\"devtools\")\n","devtools::install_github('cole-trapnell-lab/monocle3')\n","\n","#if you wish to install the development version of monocle3\n","#devtools::install_github('cole-trapnell-lab/monocle3', ref=\"develop\")"]},{"cell_type":"markdown","metadata":{"id":"KfjZD63KGPsU"},"source":["### Step: load required packages for analysis\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"7AaIUOS5GcFU"},"outputs":[],"source":["library(monocle3)\n"]},{"cell_type":"markdown","metadata":{"id":"DkNW5DCTJKYo"},"source":["In this example, we will take an example dataset"]},{"cell_type":"code","source":["expression_matrix <- readRDS(url(\"https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_expression.rds\"))\n","cell_metadata <- readRDS(url(\"https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_colData.rds\"))\n","gene_annotation <- readRDS(url(\"https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_rowData.rds\"))\n"],"metadata":{"id":"b-KvbDHuQ3D4"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Make a new cell_data_set (CDS) object as follows:\n"],"metadata":{"id":"i3j1OBq1QoVF"}},{"cell_type":"code","source":["# Make the CDS object\n","cds <- monocle3::new_cell_data_set(expression_matrix,\n","                         cell_metadata = cell_metadata,\n","                         gene_metadata = gene_annotation)"],"metadata":{"id":"X4SIB0-4Q6sv"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["VERY IMPORTANT - do NOT use `as.matrix()` on the sparse matrix object. It will convert the sparse matrix into dense matrix object . Dense matrix takes up 20 times more space than sparse matrice"],"metadata":{"id":"2piY5B7mQoEJ"}},{"cell_type":"markdown","source":["## Pre-process the data\n","This step is where you tell Monocle 3 how you want to normalize the data, whether to use Principal Components Analysis (the standard for RNA-seq) or Latent Semantic Indexing (common in ATAC-seq), and how to remove any batch effects.\n","\n","This time, we will use a different strategy for batch correction, which includes what Packer & Zhu et al did in their original analysis:\n","\n","Note: Your data will not have the loading batch information demonstrated here, you will correct batch using your own batch information. -- data is in the form of batches\n","\n","* `align_cds()` aligns groups of cells (i.e. batches)\n","* `residual_model_formula_str` is for subtracting continuous effects. You can use this to control for things like the fraction of mitochondrial reads in each cell, which is sometimes used as a QC metric for each cell.\n","\n","In this experiment (as in many scRNA-seq experiments), some cells spontanously lyse, releasing their mRNAs into the cell suspension immediately prior to loading into the single-cell library prep. This \"supernatant RNA\" contaminates each cells' transcriptome profile to a certain extent. Fortunately, it is fairly straightforward to estimate the level of background contamination in each batch of cells and subtract it, which is what Packer et al did in the original study. Each of the columns bg.300.loading, bg.400.loading, corresponds to a background signal that a cell might be contaminated with. Passing these colums as terms in the residual_model_formula_str tells align_cds() to subtract these signals prior to dimensionality reduction, clustering, and trajectory inference. Note that you can call align_cds() with alignment_group, residual_model_formula, or both."],"metadata":{"id":"Mzz60JchRJfX"}},{"cell_type":"code","source":["cds <- preprocess_cds(cds, num_dim = 50)\n","\n","cds <- align_cds(cds,\n","                 alignment_group = \"batch\",\n","                 residual_model_formula_str = \"~ bg.300.loading + bg.400.loading + bg.500.1.loading + bg.500.2.loading + bg.r17.loading + bg.b01.loading + bg.b02.loading\")"],"metadata":{"id":"VFJ0J3GrRO0D"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Reduce dimensionality and visualize the cells\n","Despite the fact that we are only looking at a small slice of this dataset, Monocle reconstructs a trajectory with numerous branches. Overlaying the manual annotations on the UMAP reveals that these branches are principally occupied by one cell type."],"metadata":{"id":"VCKjAlh3RUAS"}},{"cell_type":"code","source":["cds <- reduce_dimension(cds, reduction_method = \"UMAP\")\n","\n","plot_cells(cds,\n","           label_groups_by_cluster=FALSE,\n","           color_cells_by = \"cell.type\")"],"metadata":{"id":"zDkvgKhWRcfF"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Group cells into clusters\n","Although cells may continuously transition from one state to the next with no discrete boundary between them, Monocle does not assume that all cells in the dataset descend from a common transcriptional \"ancestor\". In many experiments, there might in fact be multiple distinct trajectories. For example, in a tissue responding to an infection, tissue resident immune cells and stromal cells will have very different initial transcriptomes, and will respond to infection quite differently, so they should be a part of the same trajectory.\n","\n","Monocle is able to learn when cells should be placed in the same trajectory as opposed to separate trajectories through its clustering procedure. Recall that we run cluster_cells(), each cell is assigned not only to a cluster but also to a partition. When you are learning trajectories, each partition will eventually become a separate trajectory. We run cluster_cells()as before."],"metadata":{"id":"NemUzRz-Rq8M"}},{"cell_type":"code","source":["cds <- cluster_cells(cds)\n"],"metadata":{"id":"m0Y1pQcsR9JP"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["IMORTANT - The cluster_cells() also divides the cells into larger, more well separated groups called **partitions**, using a statistical test from Alex Wolf et al, introduced as part of their PAGA algorithm. You can visualize these partitions like this:"],"metadata":{"id":"E3natfGvR-Lb"}},{"cell_type":"code","source":["plot_cells(cds, color_cells_by = \"partition\")\n"],"metadata":{"id":"Ef0sQUN7SIVR"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Learn the trajectory graph\n","Next, we will fit a principal graph within each partition using the learn_graph() function:"],"metadata":{"id":"r8WfZ66oWujq"}},{"cell_type":"code","source":["cds <- learn_graph(cds)\n","plot_cells(cds,\n","           color_cells_by = \"cell.type\",\n","           label_groups_by_cluster=FALSE,\n","           label_leaves=FALSE,\n","           label_branch_points=FALSE)"],"metadata":{"id":"Y6-hkzE6WxY9"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Order the cells in pseudotime\n","Once we've learned a graph, we are ready to order the cells according to their progress through the developmental program. Monocle measures this progress in pseudotime. Pseudotime is a measure of how much progress an individual cell has made through a process such as cell differentiation.\n","\n","In order to place the cells in order, we need to tell Monocle where the \"beginning\" of the biological process is. We do so by choosing regions of the graph that we mark as \"roots\" of the trajectory. In time series experiments, this can usually be accomplished by finding spots in the UMAP space that are occupied by cells from early time points:"],"metadata":{"id":"xqTeWYPXWucy"}},{"cell_type":"code","source":["plot_cells(cds,\n","           color_cells_by = \"embryo.time.bin\",\n","           label_cell_groups=FALSE,\n","           label_leaves=TRUE,\n","           label_branch_points=TRUE,\n","           graph_label_size=1.5)"],"metadata":{"id":"sJWTHp1MW2Mj"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["cds <- order_cells(cds)"],"metadata":{"id":"utduAidUW4eb"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Choose root nodes\n","In the above example, we just chose one location, but you could pick as many as you want. Plotting the cells and coloring them by pseudotime shows how they were ordered:"],"metadata":{"id":"0VnIIjqKW53e"}},{"cell_type":"code","source":["plot_cells(cds,\n","           color_cells_by = \"pseudotime\",\n","           label_cell_groups=FALSE,\n","           label_leaves=FALSE,\n","           label_branch_points=FALSE,\n","           graph_label_size=1.5)"],"metadata":{"id":"aY54Vp8fW7wh"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Some of the cells are gray. This means they have infinite pseudotime, because they were not reachable from the root nodes that were picked. In general, any cell on a partition that lacks a root node will be assigned an infinite pseudotime. In general, you should choose at least one root per partition.\n","\n","It's often desirable to specify the root of the trajectory programmatically, rather than manually picking it. The function below does so by first grouping the cells according to which trajectory graph node they are nearest to. Then, it calculates what fraction of the cells at each node come from the earliest time point. Then it picks the node that is most heavily occupied by early cells and returns that as the root."],"metadata":{"id":"MjBQGSynW9vC"}},{"cell_type":"code","source":["# a helper function to identify the root principal points:\n","get_earliest_principal_node <- function(cds, time_bin=\"130-170\"){\n","  cell_ids <- which(colData(cds)[, \"embryo.time.bin\"] == time_bin)\n","\n","  closest_vertex <-\n","  cds@principal_graph_aux[[\"UMAP\"]]$pr_graph_cell_proj_closest_vertex\n","  closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])\n","  root_pr_nodes <-\n","  igraph::V(principal_graph(cds)[[\"UMAP\"]])$name[as.numeric(names\n","  (which.max(table(closest_vertex[cell_ids,]))))]\n","\n","  root_pr_nodes\n","}\n","cds <- order_cells(cds, root_pr_nodes=get_earliest_principal_node(cds))\n","\n","#Passing the programatically selected root node to order_cells() via the root_pr_nodeargument yields:\n","\n","plot_cells(cds,\n","           color_cells_by = \"pseudotime\",\n","           label_cell_groups=FALSE,\n","           label_leaves=FALSE,\n","           label_branch_points=FALSE,\n","           graph_label_size=1.5)"],"metadata":{"id":"-XV185ldXAv5"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Subset cells by branch\n","It is often useful to subset cells based on their branch in the trajectory. The function choose_graph_segments allows you to do so interactively."],"metadata":{"id":"aQkKMCOlXCo8"}},{"cell_type":"code","source":["cds_sub <- choose_graph_segments(cds)\n"],"metadata":{"id":"vqHUbr38XD97"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Working with 3D trajectories\n"],"metadata":{"id":"K-hbq02kXFdr"}},{"cell_type":"code","source":["cds_3d <- reduce_dimension(cds, max_components = 3)\n","cds_3d <- cluster_cells(cds_3d)\n","cds_3d <- learn_graph(cds_3d)\n","cds_3d <- order_cells(cds_3d, root_pr_nodes=get_earliest_principal_node(cds))\n","\n","cds_3d_plot_obj <- plot_cells_3d(cds_3d, color_cells_by=\"partition\")"],"metadata":{"id":"dRyh4NS-XHLK"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["This is the end of this tutorial"],"metadata":{"id":"hEwAVLixR902"}},{"cell_type":"markdown","source":["## References\n","* Monocle documentation: https://cole-trapnell-lab.github.io/monocle3/docs/getting_started/\n","* Galaxy training: https://training.galaxyproject.org/training-material/topics/single-cell/tutorials/scrna-case_monocle3-rstudio/tutorial.html#monocle-workflow"],"metadata":{"id":"82lFuPOZUhEj"}}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"R","name":"ir"},"language_info":{"name":"R"}},"nbformat":4,"nbformat_minor":0}