{"cells":[{"cell_type":"markdown","metadata":{"id":"9ltdwXvN0dio"},"source":["# scRNA-seq Monocle Tutorial 3 - Differential expression\n","\n","Authors: Krithika Bhuvaneshwar, Yuriy Gusev\n","\n","Affiliation: Innovation Center For Biomedical Informatics (ICBI), and Biomedical Informatics Shared Resource (BISR) at Georgetown University Medical Center (GUMC)\n","\n","***More about our research work:***\n","* *ICBI: https://icbi.georgetown.edu*\n","* *BISR: https://icbi.georgetown.edu/bisr/ and https://lombardi.georgetown.edu/research/sharedresources/bbsr/*\n"]},{"cell_type":"markdown","metadata":{"id":"9KMphPwYFegF"},"source":["##  Before we start\n","\n","If using google colab, change the colab runtype environment to R (default is python). Go to  Runtime -> Change runtime type -> In the \"Notebook settings\", change environment to \"R\""]},{"cell_type":"markdown","metadata":{"id":"JxM7MTI0GD9U"},"source":["## Monocle software\n","Monocle software: https://cole-trapnell-lab.github.io/monocle3/docs/installation/\n","\n","* Monocle 3 is designed for use with absolute transcript counts (e.g. from UMI experiments).\n","* Monocle 3 works \"out-of-the-box\" with the transcript count matrices produced by Cell Ranger, the software pipeline for analyzing experiments from the 10X Genomics Chromium instrument.\n","* Monocle 3 also works well with data from other RNA-Seq workflows such as sci-RNA-Seq and instruments like the Biorad ddSEQ.\n","\n","Reference:\n","Reference: https://cole-trapnell-lab.github.io/monocle3/docs/differential/"]},{"cell_type":"markdown","metadata":{"id":"_Bur2rTaE5gb"},"source":["### Install monocle package - this can take a SEVERAL minutes\n","Note - if you encounter errors during installaton, please refer this page: https://cole-trapnell-lab.github.io/monocle3/docs/installation/"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"rpgabL0TEXdT"},"outputs":[],"source":["#install monocle3 through the cole-trapnell-lab GitHub\n","install.packages(\"devtools\")\n","devtools::install_github('cole-trapnell-lab/monocle3')\n","\n","#if you wish to install the development version of monocle3\n","#devtools::install_github('cole-trapnell-lab/monocle3', ref=\"develop\")"]},{"cell_type":"markdown","metadata":{"id":"KfjZD63KGPsU"},"source":["### Step: load required packages for analysis\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"7AaIUOS5GcFU"},"outputs":[],"source":["library(monocle3)\n"]},{"cell_type":"markdown","metadata":{"id":"DkNW5DCTJKYo"},"source":["In this example, we will take an example dataset"]},{"cell_type":"code","source":["expression_matrix <- readRDS(url(\"https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_expression.rds\"))\n","cell_metadata <- readRDS(url(\"https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_colData.rds\"))\n","gene_annotation <- readRDS(url(\"https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_rowData.rds\"))\n"],"metadata":{"id":"b-KvbDHuQ3D4"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Make a new cell_data_set (CDS) object as follows:\n"],"metadata":{"id":"i3j1OBq1QoVF"}},{"cell_type":"code","source":["# Make the CDS object\n","cds <- monocle3::new_cell_data_set(expression_matrix,\n","                         cell_metadata = cell_metadata,\n","                         gene_metadata = gene_annotation)"],"metadata":{"id":"X4SIB0-4Q6sv"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Performing differential expression analysis on all genes in a cell_data_set object can take anywhere from minutes to hours, depending on how complex the analysis is.\n","\n","Begin with a small set of genes that we know are important in ciliated neurons to demonstrate Monocle's capabilities:"],"metadata":{"id":"PwsTU5pIX5kN"}},{"cell_type":"code","source":["\n","ciliated_genes <- c(\"che-1\",\n","                    \"hlh-17\",\n","                    \"nhr-6\",\n","                    \"dmd-6\",\n","                    \"ceh-36\",\n","                    \"ham-1\")\n","\n","#Make a new cell_data_set (CDS) object as follows:\n","\n","cds_subset <- cds[rowData(cds)$gene_short_name %in% ciliated_genes,]\n","\n","dim(cds_subset)\n"],"metadata":{"id":"2LP-eExaX9-F"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["VERY IMPORTANT - do NOT use `as.matrix()` on the sparse matrix object. It will convert the sparse matrix into dense matrix object . Dense matrix takes up 20 times more space than sparse matrice"],"metadata":{"id":"2piY5B7mQoEJ"}},{"cell_type":"markdown","source":["## Regression model details\n","Monocle works by fitting a regression model to each gene. You can specify this model to account for various factors in your experiment (time, treatment, and so on).\n","\n","`gene_fits` is a tibble that contains a row for each gene. The `model` column contains generalized linear model objects, each of which aims to explain the expression of a gene across the cells using the equation above. The parameter `model_formula_str` should be a string specifying the model formula. The model formulae you use in your tests can include any term that exists as a column in the colData table,"],"metadata":{"id":"aC9-CVuYYCWx"}},{"cell_type":"code","source":["gene_fits <- fit_models(cds_subset,\n","                        model_formula_str = \"~embryo.time\")\n","\n","gene_fits"],"metadata":{"id":"MxxSve0KYGQu"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Find out Which of these genes have time-dependent expression. First, we extract a table of coefficients from each model using the coefficient_table() function"],"metadata":{"id":"FHaZh-ywYN1k"}},{"cell_type":"code","source":["fit_coefs <- coefficient_table(gene_fits)\n","fit_coefs"],"metadata":{"id":"rA4HKbEKYSU9"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Extract the time terms:\n"],"metadata":{"id":"khafQvz1YNvm"}},{"cell_type":"code","source":["emb_time_terms <- fit_coefs %>% filter(term == \"embryo.time\")\n"],"metadata":{"id":"T6lNDkllYWRA"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Pull out the genes that have a significant time component. coefficient_table() tests whether each coefficient differs significantly from zero under the Wald test. By default, coefficient_table() adjusts these p-values for multiple hypothesis testing using the method of Benjamini and Hochberg. These adjusted values can be found in the q_value column. We can filter the results and control the false discovery rate as follows:"],"metadata":{"id":"4a4ZBrr2YNq_"}},{"cell_type":"code","source":["emb_time_terms %>% filter (q_value < 0.05) %>%\n","         select(gene_short_name, term, q_value, estimate)"],"metadata":{"id":"faYMw_aXYZZi"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["We can see that five of the six genes significantly vary as a function of time.\n","\n","Monocle also provides some easy ways to plot the expression of a small set of genes grouped by the factors you use during differential analysis. This helps you visualize the differences revealed by the tests above. One type of plot is a \"violin\" plot."],"metadata":{"id":"bbIviNRqYbXc"}},{"cell_type":"code","source":["plot_genes_violin(cds_subset, group_cells_by=\"embryo.time.bin\", ncol=2) +\n","      theme(axis.text.x=element_text(angle=45, hjust=1))"],"metadata":{"id":"RC1xwlRIYc9z"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Controlling for batch effects and other factors\n"],"metadata":{"id":"jlFQKoqUYNj4"}},{"cell_type":"code","source":["gene_fits <- fit_models(cds_subset,\n","                        model_formula_str = \"~embryo.time + batch\")\n","fit_coefs <- coefficient_table(gene_fits)\n","fit_coefs %>% filter(term != \"(Intercept)\") %>%\n","      select(gene_short_name, term, q_value, estimate)"],"metadata":{"id":"RT1bG6WLYhQM"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Evaluating models of gene expression\n"],"metadata":{"id":"Fpcsp49EYiWL"}},{"cell_type":"code","source":["evaluate_fits(gene_fits)"],"metadata":{"id":"vgnKZA92Yjz0"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Choosing a distribution for modeling gene expression\n","Monocle uses generalized linear models to capture how a gene's expression depends on each variable in the experiment. These models require you to specify a distribution that describes gene expression values. Most studies that use this approach to analyze their gene expression data use the negative binomial distribution, which is often appropriate for sequencing read or UMI count data. The negative binomial is at the core of many packages for RNA-seq analysis, such as DESeq2.\n","\n","Monocle's `fit_models()` supports the\n","* negative binomial distribution\n","* Poisson,\n","* Binomial,\n","* quasipoisson\n","\n","The default is the \"quasipoisson\", which is very similar to the negative binomial. Quasipoisson is a a bit less accurate than the negative binomial but much faster to fit, making it well suited to datasets with thousands of cells."],"metadata":{"id":"_Dm-RpR9YiMr"}},{"cell_type":"markdown","source":["This is the end of this tutorial"],"metadata":{"id":"hEwAVLixR902"}},{"cell_type":"markdown","source":["## References\n","* Monocle documentation: https://cole-trapnell-lab.github.io/monocle3/docs/getting_started/\n","* Galaxy training: https://training.galaxyproject.org/training-material/topics/single-cell/tutorials/scrna-case_monocle3-rstudio/tutorial.html#monocle-workflow"],"metadata":{"id":"82lFuPOZUhEj"}}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"R","name":"ir"},"language_info":{"name":"R"}},"nbformat":4,"nbformat_minor":0}