{"cells":[{"cell_type":"markdown","metadata":{"id":"9ltdwXvN0dio"},"source":["# scRNA-seq Monocle Tutorial 1\n","\n","Authors: Krithika Bhuvaneshwar, Yuriy Gusev\n","\n","Affiliation: Innovation Center For Biomedical Informatics (ICBI), and Biomedical Informatics Shared Resource (BISR) at Georgetown University Medical Center (GUMC)\n","\n","***More about our research work:***\n","* *ICBI: https://icbi.georgetown.edu*\n","* *BISR: https://icbi.georgetown.edu/bisr/ and https://lombardi.georgetown.edu/research/sharedresources/bbsr/*\n"]},{"cell_type":"markdown","metadata":{"id":"MuA9crRB6shg"},"source":["## General Steps in analysis\n","Based on the workflow mentioned in the e-book \"Orchestrating Single-Cell Analysis with Bioconductor\"\n","* Removal of low-quality cells\n","* Normalization and log-transformation\n","* Modeling of the mean-variance trend across genes\n","* A principal components analysis on the highly variable genes\n","* Clustering with graph-based methods\n","* Dimensionality reductions (t-SNE/UMAP)\n","* Marker detection for each cluster\n","* Make custom cell selections and detect markers for this selection\n","* Cell type annotation for each cluster across user selected reference datasets\n","* Perform Integration or Batch correction using MNN correction.\n","* Support Multi-modal analysis for Cite-seq data\n","* Perform analysis on subsets (filter based on cell annotation)\n","\n","\n","\n"]},{"cell_type":"markdown","metadata":{"id":"9KMphPwYFegF"},"source":["##  Before we start\n","\n","If using google colab, change the colab runtype environment to R (default is python). Go to  Runtime -> Change runtime type -> In the \"Notebook settings\", change environment to \"R\""]},{"cell_type":"markdown","metadata":{"id":"JxM7MTI0GD9U"},"source":["## Monocle software\n","Monocle software: https://cole-trapnell-lab.github.io/monocle3/docs/installation/\n","\n","* Monocle 3 is designed for use with absolute transcript counts (e.g. from UMI experiments).\n","* Monocle 3 works \"out-of-the-box\" with the transcript count matrices produced by Cell Ranger, the software pipeline for analyzing experiments from the 10X Genomics Chromium instrument.\n","* Monocle 3 also works well with data from other RNA-Seq workflows such as sci-RNA-Seq and instruments like the Biorad ddSEQ."]},{"cell_type":"markdown","source":["### Monocle workflow\n","* Load scRNA-seq data matrix\n","* Pre-process: normalize, remove batch effects\n","* Non-linear dimensionality reduction: t-SNE, UMAP\n","* Cluster cells\n","* Compare clusters: identify top markers, targeted contrasts\n","* Trajectory analysis"],"metadata":{"id":"LHvrwtvaTlMt"}},{"cell_type":"markdown","metadata":{"id":"4mc-l9NGER1z"},"source":["## Example using Monocle software in R\n"]},{"cell_type":"markdown","metadata":{"id":"_Bur2rTaE5gb"},"source":["### Install monocle package - this can take SEVERAL minutes\n","Note - if you encounter errors during installaton, please refer this page: https://cole-trapnell-lab.github.io/monocle3/docs/installation/"]},{"cell_type":"code","execution_count":5,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":287},"id":"rpgabL0TEXdT","executionInfo":{"status":"error","timestamp":1689306704116,"user_tz":240,"elapsed":641,"user":{"displayName":"Krithika Bhuvaneshwar","userId":"06224013107515748971"}},"outputId":"6d0ac64a-f2f5-4cf3-f1e3-fdec93d0f5dc"},"outputs":[{"output_type":"stream","name":"stderr","text":["Installing package into ‘/usr/local/lib/R/site-library’\n","(as ‘lib’ is unspecified)\n","\n","Warning message:\n","“package ‘monocle3’ is not available for this version of R\n","\n","A version of this package for your version of R might be available elsewhere,\n","see the ideas at\n","https://cran.r-project.org/doc/manuals/r-patched/R-admin.html#Installing-packages”\n"]},{"output_type":"error","ename":"ERROR","evalue":"ignored","traceback":["Error in library(monocle3): there is no package called ‘monocle3’\nTraceback:\n","1. library(monocle3)"]}],"source":["#install monocle3 through the cole-trapnell-lab GitHub\n","#install.packages(\"devtools\")\n","#devtools::install_github('cole-trapnell-lab/monocle3')\n","\n","install.packages(\"monocle3\")\n","library(monocle3)"]},{"cell_type":"markdown","metadata":{"id":"KfjZD63KGPsU"},"source":["### Step: load required packages for analysis\n"]},{"cell_type":"code","execution_count":4,"metadata":{"id":"7AaIUOS5GcFU","colab":{"base_uri":"https://localhost:8080/","height":131},"executionInfo":{"status":"error","timestamp":1689306416346,"user_tz":240,"elapsed":14,"user":{"displayName":"Krithika Bhuvaneshwar","userId":"06224013107515748971"}},"outputId":"00bdfa64-2330-4612-83c6-5c3e383bd1e8"},"outputs":[{"output_type":"error","ename":"ERROR","evalue":"ignored","traceback":["Error in library(monocle3): there is no package called ‘monocle3’\nTraceback:\n","1. library(monocle3)"]}],"source":["library(monocle3)\n","\n","# The tutorial shown below and on subsequent pages uses two additional packages:\n","library(ggplot2)\n","library(dplyr)"]},{"cell_type":"markdown","source":["## Introduction\n","Once the raw data in FASTQ file is processed, the data files can be in several formats\n","\n","* Data can be MatrixMarket (MTX) format The MTX format is a sparse matrix format with genes on the rows and cells on the columns as output by Cell Ranger which is the output from the 10X machine, along with two metadata files (features information and cell information).\n","* RDS format Includes 3 files - expression_matrix, cell_metadata, gene_metadata (gene_annotation) files\n","\n","Let us assume the input data sparse matrix is in a RDS format"],"metadata":{"id":"IHqRQisqQwLQ"}},{"cell_type":"markdown","metadata":{"id":"vFb6ICp7GhUD"},"source":["### Step: Load input data\n","* Input data format : cell by gene expression matrix. The object name is `cell_data_set` class. This is dervied from  Bioconductor's `SingleCellExperiment` class. It requires three input files:\n","* `expression_matrix`, a numeric matrix of expression values, where rows are genes, and columns are cells\n","* `cell_metadata`, a data frame, where rows are cells, and columns are cell attributes (such as cell type, culture condition, day captured, etc.)\n","* `gene_metadata`, an data frame, where rows are features (e.g. genes), and columns are gene attributes, such as biotype, gc content, etc.\n","\n","Notes:\n","* The expression value matrix must:\n","(a) have the same number of columns as the cell_metadata has rows.\n","(b) have the same number of rows as the gene_metadata has rows\n","(c) row names of the cell_metadata object should match the column names of the expression matrix.\n","(d) row names of the gene_metadata object should match row names of the expression matrix.\n","(e) one of the columns of the gene_metadata should be named \"gene_short_name\", which represents the gene symbol or simple name (generally used for plotting) for each gene.\n","\n"]},{"cell_type":"markdown","metadata":{"id":"DkNW5DCTJKYo"},"source":["In this example, we will take an example dataset"]},{"cell_type":"code","source":["expression_matrix <- readRDS(url(\"https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/cao_l2_expression.rds\"))\n","\n","cell_metadata <- readRDS(url(\"https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/cao_l2_colData.rds\"))\n","\n","gene_annotation <- readRDS(url(\"https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/cao_l2_rowData.rds\"))"],"metadata":{"id":"b-KvbDHuQ3D4"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Make a new cell_data_set (CDS) object as follows:\n"],"metadata":{"id":"i3j1OBq1QoVF"}},{"cell_type":"code","source":["# Make the CDS object\n","cds <- monocle3::new_cell_data_set(expression_matrix,\n","                         cell_metadata = cell_metadata,\n","                         gene_metadata = gene_annotation)"],"metadata":{"id":"X4SIB0-4Q6sv"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Note - for data from 10X this is how data can be imprted\n","```{r, eval=FALSE}\n","# Provide the path to the Cell Ranger output.\n","cds <- load_cellranger_data(\"input_pbmc3k/filtered_gene_bc_matrices/hg19/\")\n","\n","#OR\n","cds <- load_mm_data(mat_path = \"input_pbmc3k/filtered_gene_bc_matrices/hg19/matrix.mtx\",\n","                    feature_anno_path = \"input_pbmc3k/filtered_gene_bc_matrices/hg19/genes.tsv\",\n","                    cell_anno_path = \"input_pbmc3k/filtered_gene_bc_matrices/hg19/barcodes.tsv\")\n","\n","#To work with your data in a sparse format, simply provide it to Monocle 3 as a sparse matrix from the Matrix package:\n","\n","cds <- new_cell_data_set(as(umi_matrix, \"sparseMatrix\"),\n","cell_metadata = cell_metadata,\n","gene_metadata = gene_metadata)\n","```"],"metadata":{"id":"exns2KuBQoKv"}},{"cell_type":"markdown","source":["VERY IMPORTANT - do NOT use `as.matrix()` on the sparse matrix object. It will convert the sparse matrix into dense matrix object . Dense matrix takes up 20 times more space than sparse matrice"],"metadata":{"id":"2piY5B7mQoEJ"}},{"cell_type":"markdown","source":["## Pre-process the data\n","This step is where you tell Monocle 3 how you want to normalize the data, whether to use Principal Components Analysis (the standard for RNA-seq) or Latent Semantic Indexing (common in ATAC-seq), and how to remove any batch effects. We will just use the standard PCA method in this demonstration. When using PCA, you should specify the number of principal components you want Monocle to compute.\n","\n","It's a good idea to check that you're using enough PCs to capture most of the variation in gene expression across all the cells in the data set. You can look at the fraction of variation explained by each PC using `plot_pc_variance_explained():`\n","\n","We can see that using more than 100 PCs would capture only a small amount of additional variation, and each additional PC makes downstream steps in Monocle slower.\n","\n","### IMPORTANT -\n","According to Galaxy documentation, The plot shows that actually using more than ~100 PCs captures only a small amount of additional variation. However, if we look at how the cells are plotted on 2D graph when using different values of PCs, it is easier to visualise how the num_dim actually affects the output.\n","In the Galaxy example, 4 plots were done : 145 PCs, 165 PCs, 200 PCs, 210 PCs,\n","The graph is different each time.\n","Galaxy example used a value of of 210, which made the most sense for their dataset."],"metadata":{"id":"Mzz60JchRJfX"}},{"cell_type":"code","source":["cds <- preprocess_cds(cds, num_dim = 100)\n","\n","plot_pc_variance_explained(cds)"],"metadata":{"id":"VFJ0J3GrRO0D"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Reduce dimensionality and visualize the cells\n","Now we're ready to visualize the cells. To do so, you can use either t-SNE, which is very popular in single-cell RNA-seq, or UMAP, which is increasingly common.\n","\n","### t-SNE vs UMAP\n","The main difference between t-SNE and UMAP is the interpretation of the distance between objects or \"clusters\"\n","* Both t-SNE and UMAP are non-linear reductions\n","* UMAP operates on the k-nearest neighbours graph (for some small value of k), exactly as t-SNE does\n","* t-SNE video https://www.youtube.com/watch?v=NEaUSP4YerM\n","* t-SNE preserves local structure in the data.This means with t-SNE you cannot interpret the distance between clusters A and B at different ends of your plot. You cannot infer that these clusters are more dissimilar than A and C, where C is closer to A in the plot. But within cluster A, you can say that points close to each other are more similar objects than points at different ends of the cluster image.\n","*  It captures global structure better than t-SNE. With UMAP, you should be able to interpret both the distances between / positions of points and clusters.\n","* Both algorithms are highly stochastic and very much dependent on choice of hyperparameters (t-SNE even more than UMAP) and can yield very different results in different runs,\n","* while t-SNE doesn't have much use outside of visualization, UMAP is a general-purpose dimensionality reduction technique that can be used as preprocessing for machine learning.\n","\n","Monocle 3 uses UMAP by default, as it is both faster and better suited for clustering and trajectory analysis in RNA-seq. To reduce the dimensionality of the data down into the X, Y plane so we can plot it easily, call reduce_dimension():\n","\n","If you have a relatively large dataset (with >10,000 cells or more), you may want to take advantage of options that can accelerate UMAP. Passing umap.fast_sgd=TRUE to reduce_dimension() will use a fast stochastic gradient descent method inside of UMAP. If your computer has multiple cores, you can use the cores argument to make UMAP multithreaded. However, invoking reduce_dimension() with either of these options will make it produce slighly different output each time you run it. If this is acceptable to you, you could see signifant reductions in the running time of reduction_dimension()."],"metadata":{"id":"VCKjAlh3RUAS"}},{"cell_type":"code","source":["cds2 <- reduce_dimension(cds,\n","                         reduction_method = \"UMAP\")\n","plot_cells(cds2)"],"metadata":{"id":"zDkvgKhWRcfF"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Each point in the plot above represents a different cell in the cell_data_set object cds. As you can see the cells form many groups, some with thousands of cells, some with only a few. Cao & Packer annotated each cell according to type manually by looking at which genes it expresses.\n","We can color the cells in the UMAP plot by the authors' original annotations using the color_cells_by argument to `plot_cells()`"],"metadata":{"id":"puTAGKVHRJX1"}},{"cell_type":"code","source":["plot_cells(cds2, color_cells_by=\"cao_cell_type\")"],"metadata":{"id":"S0kyJlbgRiuV"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Alternate option t SNE\n","fI you want, you can also use t-SNE to visualize your data. First, call reduce_dimension with reduction_method=\"tSNE\".\n","```{r}\n","cds3 <- reduce_dimension(cds2, reduction_method=\"tSNE\")\n","\n","#Then, when you call plot_cells(), pass reduction_method=\"tSNE\" to it as well:\n","plot_cells(cds3, reduction_method=\"tSNE\", color_cells_by=\"cao_cell_type\")\n","```"],"metadata":{"id":"6xGhMq0LRI7h"}},{"cell_type":"markdown","source":["## Check for and remove batch effects\n","\n","If the dataset  comprises data from MULTIPLE samples, there will be a risk that batch effects will impact analysis. So always check for batch effects when you perform dimensionality reduction.\n","\n","You should add a column to the `colData` that encodes which batch each cell is from. Then you can simply color the cells by batch. Cao & Packer et al included a \"plate\" annotation in their data, which specifies which sci-RNA-seq plate each cell originated from. Coloring the UMAP by plate reveals:"],"metadata":{"id":"urjGwvK_Rrcb"}},{"cell_type":"code","source":["plot_cells(cds2, color_cells_by=\"plate\", label_cell_groups=FALSE)\n"],"metadata":{"id":"HS36BftORto_"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Remove batch effect is by running the align_cds() function:\n"],"metadata":{"id":"fGQ3oHwLRrTo"}},{"cell_type":"code","source":["cds4 <- align_cds(cds2,\n","                 num_dim = 50,\n","                 alignment_group = \"plate\")\n","cds5 <- reduce_dimension(cds4)\n","plot_cells(cds5, color_cells_by=\"plate\", label_cell_groups=FALSE)"],"metadata":{"id":"Nt1W8Uj2R3Nf"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Group cells into clusters\n","Grouping cells into clusters is an important step in identifying the cell types represented in your data. Monocle uses a technique called **community detection** to group cells.\n","\n","This approach was introduced by Levine et al as part of the phenoGraph algorithm.\n","\n","### IMPORTANT - Resolution paramter\n","* When using standard igraph louvain clustering, the value of resolution parameter is by default set to NULL, which means that it is determined automatically. Although the resulting clusters are OK, it would be nice to get some more granularity to identify cell types more specifically.\n","* The higher the resolution value, the more clusters we get.\n","* In Galaxy example, the resolution value was set as 0.0002.\n","* In this example, the resolution value was set as 1e-5.\n","\n","You can cluster your cells using the cluster_cells() function, like this:"],"metadata":{"id":"NemUzRz-Rq8M"}},{"cell_type":"code","source":["cds6 <- cluster_cells(cds5,\n","                      resolution=1e-5)\n","plot_cells(cds6)"],"metadata":{"id":"m0Y1pQcsR9JP"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["IMORTANT - The cluster_cells() also divides the cells into larger, more well separated groups called **partitions**, using a statistical test from Alex Wolf et al, introduced as part of their PAGA algorithm. You can visualize these partitions like this:"],"metadata":{"id":"E3natfGvR-Lb"}},{"cell_type":"code","source":["plot_cells(cds6, color_cells_by=\"partition\", group_cells_by=\"partition\")\n"],"metadata":{"id":"Ef0sQUN7SIVR"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Find marker genes expressed by each cluster\n","Once cells have been clustered, we can ask what genes makes them different from one another. To do that, start by calling the top_markers() function:"],"metadata":{"id":"0T_8VRjXR-Bh"}},{"cell_type":"code","source":["marker_test_res <- top_markers(cds5,\n","                               group_cells_by=\"partition\",\n","                               reference_cells=1000,\n","                               cores=8)"],"metadata":{"id":"ACFdmUWkSMd8"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["The data frame marker_test_res contains a number of metrics for how specifically expressed each gene is in each partition. We could group the cells according to cluster, partition, or any categorical variable in colData(cds). You can rank the table according to one or more of the specificity metrics and take the top gene for each cluster. For example, pseudo_R2 is one such measure. We can rank markers according to pseudo_R2 like this:"],"metadata":{"id":"xbX8YqWXSOit"}},{"cell_type":"code","source":["top_specific_markers <- marker_test_res %>%\n","                            filter(fraction_expressing >= 0.10) %>%\n","                            group_by(cell_group) %>%\n","                            top_n(1, pseudo_R2)\n","\n","top_specific_marker_ids <- unique(top_specific_markers %>% pull(gene_id))\n","head(top_specific_marker_ids)"],"metadata":{"id":"hAjHYBanSQS5"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Now, we can plot the expression and fraction of cells that express each marker in each group with the plot_genes_by_group function:"],"metadata":{"id":"MOdf3nSVSRsi"}},{"cell_type":"code","source":["plot_genes_by_group(cds,\n","                    top_specific_marker_ids,\n","                    group_cells_by=\"partition\",\n","                    ordering_type=\"maximal_on_diag\",\n","                    max.size=3)"],"metadata":{"id":"4Tw2m1l_U-aW"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["It's often informative to look at more than one marker, which you can do just by changing the first argument to top_n():"],"metadata":{"id":"r7PMFGgESRkR"}},{"cell_type":"code","source":["top_specific_markers <- marker_test_res %>%\n","                            filter(fraction_expressing >= 0.10) %>%\n","                            group_by(cell_group) %>%\n","                            top_n(3, pseudo_R2)\n","\n","top_specific_marker_ids <- unique(top_specific_markers %>% pull(gene_id))\n","\n","plot_genes_by_group(cds,\n","                    top_specific_marker_ids,\n","                    group_cells_by=\"partition\",\n","                    ordering_type=\"cluster_row_col\",\n","                    max.size=3)"],"metadata":{"id":"tKb5CJW0VCb0"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["This is the end of this tutorial"],"metadata":{"id":"hEwAVLixR902"}},{"cell_type":"markdown","source":["## References\n","* Monocle documentation: https://cole-trapnell-lab.github.io/monocle3/docs/getting_started/\n","* Galaxy training: https://training.galaxyproject.org/training-material/topics/single-cell/tutorials/scrna-case_monocle3-rstudio/tutorial.html#monocle-workflow"],"metadata":{"id":"82lFuPOZUhEj"}}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"R","name":"ir"},"language_info":{"name":"R"}},"nbformat":4,"nbformat_minor":0}